### PR TITLE
Initial skeleton for Winch

### PR DESCRIPTION
**Pull request by saulecabrera**
This change introduces the initial skeleton for Winch, the "baseline" compiler, discussed https://github.com/bytecodealliance/rfcs/pull/28

The skeleton contains the just the necessary code for the main abstractions to support the following:

- Function prologue and epilogue (excluding stack checks)
- Calculation of local slots
- An initial pass on a generic MacroAssembler interface
- An initial pass on a x64 Assembler
- A simple CLI tool to compile WebAssembly programs (`crates/winch/src/main.rs`)
- String based emission for programs that use the following WebAssembly instructions: `i32.add`, `local.get`, `local.set`

This change doesn't fully integrate Winch with Wasmtime yet. 

**Overview**

As of this change, the high-level structure of the compiler is as follows:

```
                                                                                                                                                                                                                                                                                                                                                      
       +--------------------------------------+                                                                                                                                                                                                                                                                                                      
       |                                      |                                                                                                                                                                                                                                                                                                      
       |                                      |                                                                                                                                                                                                                                                                                                      
       |    +-----------------------+         |                                                                                                                                                                                                                                                                                                      
       |    |                       |         |                                                                                                                                                                                                                                                                                                      
       |    |  +-----------+      +--------+  |                                                                                                                                                                                                                                                                                                      
       |    |  | Compiler  | ---> |  ISA   | -+                                                                                                                                                                                                                                                                                                      
       |    |  +-----------+      +--------+                                                                                                                                                                                                                                                                                                         
       |    |                       |                                                                                                                                                                                                                                                                                                                
       |    |                       |                                                                                                                                                                                                                                                                                                                
       |    |                       v                                                                                                                                                                                                                                                                                                                
       |    |  +-----------+-----+---------+     +---------------------+     +----------------+     +-------+----------------+-------+                                                                                                                                                                                                               
       |    +> | Registers | ABI | CodeGen | --> | Register Allocation | --> | CodeGenContext | --> | Frame | MacroAssembler | Stack |                                                                                                                                                                                                               
       |       +-----------+-----+---------+     +---------------------+     +----------------+     +-------+----------------+-------+                                                                                                                                                                                                               
       |                      ^     |                                          ^                                                                                                                                                                                                                                                                     
       +----------------------+     +------------------------------------------+                                             
```
--- 

**Next steps**

As mentioned above, this change doesn't use any of Cranelift's backends for code emission, yet. These are the immediate next steps after landing this change:

- Extract the relevant Cranelift pieces into the `cranelift_asm` crate, which will be used by Winch
- Swap the string based implementation of the x64 assembler and use the assembler provided by `cranelift_asm`
- Add initial support for arm64 through `cranelift_asm`
- Improve error handling
- Increase test coverage, and include tests in CI. Opted to not do that in this change since the emission mechanism will change, so it might be easier to wait and add verification tests once we are using Cranelift's backends. 
- There are a handful of `TODO`s in the code that are expected to be addressed in the next iterations. I left them there as a reminder mostly; but if preferred they could be removed and placed into issues instead. 

--- 
**Misc** 

Winch's CLI can be used by running `cargo run -- path/to/file.wat --target=<target>`

Here's a set of WebAssembly programs and their corresponding x64 output:
<details>
<summary>Basic i32 add</summary>

```wat
(module
  (export "main" (func $main))

  (func $main (result i32)
    (i32.const 10)
    (i32.const 20)
    i32.add)
)

```

```asm
push rbp
mov rbp, rsp
mov eax, 10
add eax, 20
pop rbp
ret
```
</details>

<details>
<summary> i32 add with local operators </summary>

```wat
(module
  (export "main" (func $main))

  (func $main (result i32)
    (local $foo i32)  
    (local $bar i32)
    (i32.const 10)
    (local.set $foo)
    (i32.const 20)
    (local.set $bar)
    (local.get $foo)
    (local.get $bar)
    i32.add 
  )
)
```

```asm
push rbp
mov rbp, rsp
sub rsp, 8
mov qword [rsp], 0
mov eax, 10
mov [rsp + 4], eax
mov ecx, 20
mov [rsp], ecx
mov eax, [rsp]
mov edx, [rsp + 4]
add edx, eax
mov rax, rdx
add rsp, 8
pop rbp
ret
```
</details>

<details>
<summary>Basic i32 with function arguments </summary>

```wat
(module
  (export "main" (func $main))

  (func $main (param i32) (param i32) (result i32)
      (local.get 0)
      (local.get 1)
      i32.add
    )
)

```

```asm
push rbp
mov rbp, rsp
sub rsp, 8
mov [rsp + 4], edi
mov [rsp], esi
mov eax, [rsp]
mov ecx, [rsp + 4]
add ecx, eax
mov rax, rcx
add rsp, 8
pop rbp
ret
```
</details>


<!--

Please ensure that the following steps are all taken care of before submitting
the PR.

- [ ] This has been discussed in issue #..., or if not, please tell us why
  here.
- [ ] A short description of what this does, why it is needed; if the
  description becomes long, the matter should probably be discussed in an issue
  first.
- [ ] This PR contains test cases, if meaningful.
- [ ] A reviewer from the core maintainer team has been assigned for this PR.
  If you don't know who could review this, please indicate so. The list of
  suggested reviewers on the right can help you.

Please ensure all communication adheres to the [code of
conduct](https://github.com/bytecodealliance/wasmtime/blob/master/CODE_OF_CONDUCT.md).
-->

### CODE REVIEW

**Comment by fitzgen** [comment]
Excited for this!

File: crates/winch/Cargo.toml
Comment (bjorn3) [inline crates/winch/Cargo.toml:14]: This should probably replaced with features on winch that enable the respective feature of cranelift-codegen and additionally enabling the target architecture if no architecture is selected, just like for cranelift-codegen.

File: crates/winch/Cargo.toml
Comment (cfallin) [inline crates/winch/Cargo.toml:2]: I think there are two different ways we could go here: a `winch` crate in its own subdirectory, as `cranelift` is, or a `wasmtime-winch` crate in `crates/winch` with a version number synchronized with the rest of the Wasmtime crates. Given the dependency on `wasmtime-environ` below I think it's probably more properly considered part of the Wasmtime family of crates, but if eventually factor that out, then it at least in theory could be its own thing usable separately (in the way that Cranelift is).

but tl;dr, a crate in `crates/$NAME` is part of Wasmtime and is named as `wasmtime-$NAME`...

File: crates/winch/Cargo.toml
Comment (cfallin) [inline crates/winch/Cargo.toml:14]: Comment here too that this is a temporary dependency?

File: crates/winch/src/abi/mod.rs
Comment (cfallin) [inline crates/winch/src/abi/mod.rs:11]: I'll note here but applies generally: overall in doc-comments we've tried to stick to capitalized-with-periods-at-end style; many comments here don't end with a `.`.

File: crates/winch/src/codegen.rs
Comment (cfallin) [inline crates/winch/src/codegen.rs:61]: `word_size` should probably be obtained from the `MacroAssembler` or other platform-specific trait?

File: crates/winch/src/codegen.rs
Comment (cfallin) [inline crates/winch/src/codegen.rs:127]: Can we factor out this logic into a `memset` helper on the `MacroAssembler`, that either unrolls a short sequence of stores or calls a helper?

File: crates/winch/src/compiler/mod.rs
Comment (cfallin) [inline crates/winch/src/compiler/mod.rs:19]: If we want to break the dependency on Wasmtime, the way to do that would be to have a `crates/winch` crate analogous to `crates/cranelift` (`wasmtime-cranelift`) that pulls in the actual Winch toplevel crate (just `winch`, or `winch-codegen` or somesuch) and impls this Wasmtime trait. We can always refactor in that direction later, but if you wanted to do it now I think that could be reasonable as well.

File: crates/winch/src/frame/mod.rs
Comment (cfallin) [inline crates/winch/src/frame/mod.rs:73]: References to SpiderMonkey are fine here but for the benefit of readers later, links to the source (via searchfox) might be helpful -- so e.g. starting out with "This algorithm is similar to SpiderMonkey's frame layout logic in LINK... The main differences are: ..."

File: crates/winch/src/isa/aarch64/abi.rs
Comment (cfallin) [inline crates/winch/src/isa/aarch64/abi.rs:68]: We should assert here that there is only one result type, right? (I think that wasmparser probably accepts multi-value Wasm by default now)

Doing the fully general thing is fine to defer to later, but we'll need to support it at some point; in Cranelift our ABI uses the arg regs (x0-x7 on aarch64) as eight returns and then has a "return value area" on the stack for more than that...

Comment (cfallin) [reply crates/winch/src/isa/aarch64/abi.rs:68]: (Ah, reading further below to the x64 version, I see an assert there, so I think it's just missing here)

File: crates/winch/src/visitor.rs
Comment (cfallin) [inline crates/winch/src/visitor.rs:649]: Ah, now I understand why two-arg `add` is kind of OK actually: Wasm `add` (and other binops) always consume their args, so it's fine to overwrite `dst`. That actually makes a lot of sense to me, so maybe it's fine to keep two-arg forms of the masm methods, unless it later hinders lowerings of e.g. SIMD ops or address computations or ... that use temps.

File: crates/winch/src/visitor.rs
Comment (cfallin) [inline crates/winch/src/visitor.rs:23]: I'm wondering if it's possible to avoid the complexity of these macros -- I don't like them for several reasons:

* It duplicates knowledge here (the whole set of Wasm opcodes) unnecessarily;
* It's very unreadable;
* The dispatching mechanism feels somewhat "magic" as a result, in a bad way (I can't trace what's going on).

I think I'd prefer a toplevel translation driver loop that's something like:

```rust
for op in operators {
  match op {
    Operator::Add => self.visit_add();
    ...
  }
}
```

maybe with that `match` in a handwritten `fn visit()`. In other words, just expand out what this macro would generate by hand.

Or said another way: it seems these macros are only used once. Given that, why not expand them in the place that they're used, to reduce complexity?

**Review by cfallin — COMMENTED** [review]
Thanks @saulecabrera for all the hard work in bootstrapping this -- I'm excited to see it in this shape and looking forward to the rest!

I have a bunch of comments below from my readthrough, but hopefully they aren't too burdensome; in general I (still) think it makes sense to iterate on this in-tree, so all the partially-done and stubbed-out stuff is totally fine, and we can refactor other things as needed later too.

**Comment by saulecabrera** [comment]
Thanks for the feedback @cfallin; I'll go through each of the review items separately; at a high-level I think that I'd like to tackle some of the refactoring items in this PR (e.g. moving winch to a top level directory and using that in `wasmtime-winch`) to have a common base moving forward, it shouldn't be too complicated and hopefully the rebasing won't be too bad either.

File: crates/winch/src/visitor.rs
Comment (saulecabrera) [reply crates/winch/src/visitor.rs]: > I'm wondering if it's possible to avoid the complexity of these macros

Yeah, definitely possible. I've addressed this in https://github.com/bytecodealliance/wasmtime/pull/4907/commits/147f901ea630bb39b2a59fa956293c6c232f4e41

FWIW, I like the visitor pattern recently introduced in `wasmparser`,  but when I was exploring introducing it I couldn't find a cleaner way to manage the support/unsupported operator sets, which led to duplication. I'd still like to keep the implementation of the `visit` function and each of the emitters in the `visit.rs` module though; I expect that this file is going to get very big as more operators are added.

File: crates/winch/src/codegen.rs
Comment (saulecabrera) [reply crates/winch/src/codegen.rs:130]: I think it's reasonable to do this since this is common for all ISAs and will be cleaner if we put this in the masm. A few clarification questions: 

1. With this logic you mean the entire piece to zero out a locals range?
2. By helper, is your idea to have this in the masm module, but not necessarily part of the trait?

File: crates/winch/src/visitor.rs
Comment (saulecabrera) [reply crates/winch/src/visitor.rs]: Thanks for chiming in here @alexcrichton; I wasn't aware of the speed benefits; I wrongly assumed the change was mostly around ergonomics. But taking a closer look at the details and [some of the comments](https://github.com/bytecodealliance/wasm-tools/pull/697#issuecomment-1210517471), your comment makes sense to me. 

I'll try out your suggestion, hopefully it'll be less complex that my previous attempt.

Comment (alexcrichton) [reply crates/winch/src/visitor.rs]: I do realize that macros are nontrivial to work with for sure though and I'm still not 100% satisfied with where `wasmparser` ended up. If you'd like some help I'm happy to assist, or additionally at this point getting something working is probably more prudent than the highest levels of optimization, so I'm happy to try to come work on this small part as a follow-up if you'd prefer too

Comment (cfallin) [reply crates/winch/src/visitor.rs]: > Given that winch's goal is compilation speed I might caution against going this direction since the Operator-based approach to both validation and translation is known to be double-digit percentage slower when at least purely validating and would likely have similar impact to the design of winch as well.

Ah, OK, I think we're slightly talking past each other, and my hand expansion into a `match` was based on my misunderstanding of what this was generating (sorry!). My main point though wasn't about the actual matching/control-flow strategy, but about the way we generate it.

Basically my point can be reduced to: there are some macros here; each macro is invoked once; can we inline the macros' expansion at that point instead? (Since this doesn't change what eventually gets compiled into the winch crate, it should be completely performance-neutral.) It didn't seem like the redundancy that the macros reduced (method args, ..) was that significant compared to the complexity of having the macros at all. And my confusion on what they were actually doing is only slightly more evidence of (either my denseness or...) the confusion imparted by macro use :-)

Comment (saulecabrera) [reply crates/winch/src/visitor.rs]: > Basically my point can be reduced to: there are some macros here; each macro is invoked once; can we inline the macros' expansion at that point instead? 

I think that in this case, the approach would be to manually define each method expected by the visitor trait, and error out on the ones that are not supported. This satisfies the readability and the performance aspects, at the cost of having to manually define a method per operator.  I'd would still like to explore if there's a middle ground in which we are able to use macros without hindering too much the readability, (e.g. following Alex's suggestion of only having a single list of unsupported operators) and perhaps the complexity can be reduced further with comments? And, thinking a bit more, if we go with this approach, as we support more and more operators, this "unsupported" list is going to shrink over time.

File: crates/winch/Cargo.toml
Comment (saulecabrera) [reply crates/winch/Cargo.toml]: Yeah this is a good idea, I've completed this and I'm happy how things are shaping up.

File: crates/winch/src/compiler/mod.rs
Comment (saulecabrera) [reply crates/winch/src/compiler/mod.rs]: I've tackled this too, I think that it's better to do this earlier rather than later.

File: crates/winch/src/main.rs
Comment (saulecabrera) [reply crates/winch/src/main.rs]: I've tackled this as part of restructuring the multiple Winch crates, so now we have this tool in it's own directory, not related with Wasmtime at all.

ref: https://github.com/bytecodealliance/wasmtime/commit/3c6f23acb5b992fe936ec4bd9dd28f8077bd8580

File: crates/winch/src/visitor.rs
Comment (saulecabrera) [reply crates/winch/src/visitor.rs]: Had a chat with Alex (thanks for the help Alex!) in which we were able to simplify the macro. @cfallin let me know if you have further thoughts on this; I'd like to keep a macro if possible, assuming that it will temporary as more operators are supported. If not I'm happy to manually define all the Visitor traits in a follow up PR. In general I feel that this is cleaner than my previous attempt.

File: crates/winch/src/codegen.rs
Comment (cfallin) [reply crates/winch/src/codegen.rs]: Right, the entire bit to zero a range of memory; and I think it'd be pretty reasonable to have it as a trait method on `MacroAssembler` that has a default body (which does some basic lowering to either unrolled stores or a loop that will work on any architecture). That also leaves room for more efficient implementations per machine later (e.g. `rep movsb` on x86 machines with [ERMS](https://stackoverflow.com/questions/43343231/enhanced-rep-movsb-for-memcpy) where that's fast, or something using SIMD, or ...).

File: crates/winch/src/visitor.rs
Comment (cfallin) [reply crates/winch/src/visitor.rs]: OK, yeah, I'm happy with how this looks -- thank you for iterating on this!

File: crates/winch/src/codegen.rs
Comment (saulecabrera) [reply crates/winch/src/codegen.rs]: Makes sense to me; I've factored this out into a method in the trait with a default implementation. I've also made the code a bit simpler.

File: crates/winch/src/abi/mod.rs
Comment (saulecabrera) [reply crates/winch/src/abi/mod.rs]: Done, thanks! Sorry for not noticing this earlier!

File: crates/winch/src/isa/aarch64/abi.rs
Comment (saulecabrera) [reply crates/winch/src/isa/aarch64/abi.rs]: Added, thanks!

File: crates/winch/src/isa/x64/masm.rs
Comment (saulecabrera) [reply crates/winch/src/isa/x64/masm.rs]: > I recall having some sort of this conversation before, but don't remember all of our conclusions

Yeah we concluded that three-arg form is probably better and that we should go with it. My original intention was to do this in the next iteration once I add some actual codegen for `aarch64` but while doing the refactoring I took care of it.

File: crates/winch/src/masm.rs
Comment (saulecabrera) [reply crates/winch/src/masm.rs]: Agreed, I've added this, thanks!

**Comment by saulecabrera** [comment]
@cfallin I've addressed all the review items, this is ready for another pass. I decided to tackle the major refactoring items in this PR, so now the structure looks like:

- `winch/src/main`: the CLI tool (`winch-tools`)
- `winch/codegen`: all the code generation pieces (`winch-codegen`)
- `crates/winch`: the wasmtime integration (`winch-wasmtime`)

I had a brief battle with the `CI/verify-publish` job, but I think that I got the handle of the expected structure;  I'd appreciate a verification there (I was thinking maybe @alexcrichton could peek and confirm too).

**Comment by saulecabrera** [comment]
Hmm it seems that my change to address the comments formatting got lost. Looking into this.

**Comment by saulecabrera** [comment]
> Hmm it seems that my change to address the comments formatting got lost. Looking into this.

Formatting was indeed missing. It was an oversight from my end.

File: winch/codegen/src/masm.rs
Comment (cfallin) [inline winch/codegen/src/masm.rs:115]: Here we need an assert that storing just one i32 will bump the start address up to `word_size`-aligned, right? (I guess one can show this is always the case on 64-bit systems; if not 8-byte aligned, but 4-byte aligned, one can be at most 4 bytes off; but that conclusion isn't generic to word size so I'd rather at least assert it here.)

**Review by cfallin — APPROVED** [review]
This all looks reasonable -- thanks a bunch for the work on addressing comments!

I just have one thought ideally addressed with an assert in the zeroing logic below, otherwise LGTM. Good to merge once addressed.

File: winch/codegen/src/masm.rs
Comment (saulecabrera) [reply winch/codegen/src/masm.rs:115]: Fixed; and yeah good catch; definitely I had assumed what you pointed out, that it's always the case in 64-bit systems.

**Review by cfallin — APPROVED** [review]

**Pull request merged**
